### PR TITLE
using PHP like its 2023!

### PR DESCRIPTION
**Pull request by miko007**
Chelo comrade!

i elevated your code to use features, a professional PHP developer would use in 2023.

This addresses most of your concerns you had with the language during your livestream, and i feel like the diff of this PR will provide you with a simple oversight of what is possible and what the language provides nowadays, without you having to learn/lookup stuff.

As you just have written this code, you probably can easily spot the differences and draw your conclusions. Maybe you will even give PHP a second chance some time in the future ;P

### CODE REVIEW

**Comment by strager** [comment]
Does this break compatibility with older versions of PHP?

**Comment by lefuturiste** [comment]
> Does this break compatibility with older versions of PHP?

yes I think it support php >= 8.1 (because enums are introduced in php 8.1)

**Comment by miko007** [comment]
> Does this break compatibility with older versions of PHP?

yes, this is correct. follwing is a deduction of the features according to PHP version:

**constructor property promotion *(for less verbosity)*:**  
this came with PHP 8.0

**strong attribute types:**  
these are available *(with the `union` types)* since PHP 8.0

**strong parameter types:**  
these are available *(with custom `class` Types)* from PHP 7.4

**strong return types:**  
this *(especially the `void` type)* came about in PHP 7.2

**the mighty `enum`s:**  
are available since PHP 8.1

so in conclusion, the minimum version to run all my changes is **PHP 8.1**. I do not think that would be a problem, as this does not have to run as a cgi script for a webserver, so using newer PHP versions should be fine.

File: c.php [hunk 1]
```diff
@@ -312,21 +305,23 @@ function parse_function($lexer) {
...
-function literal_to_py($value) {
+function literal_to_py(string|int $value) : string {
     if (is_string($value)) {
         return "\"" . $value . "\"";
     } else {
         return (string)$value;
     }
 }
 
-foreach($func->body as &$stmt) {
```
Comment (miko007) [inline c.php:329]: PHP does copy on write. The use of the reference operator *(`&`)* is neither necessary, nor recommended, as it guarantees a copy internally in this case, so it would be bad for performance.

**Comment by KeremArdicli** [comment]
For a broader support, would not it work better to use regex to get the tokens?

**Comment by strager** [comment]
@KeremArdicli I think regex is only allowed for HTML parsers. This is a C compiler.

**Comment by KeremArdicli** [comment]
with fopen() you read the content of a file as a string hence can use regex on it. So you can tokenize whatever you like. If you are good at regex, Then you can push tokens into array and can iterate on that array for the compiler part.

the whole tokenizer will not be longer than 10 lines of codes.

**Comment by davidbuzz** [comment]
lgtm

**Comment by miko007** [comment]
> with fopen() you read the content of a file as a string hence can use regex on it. So you can tokenize whatever you like. If you are good at regex, Then you can push tokens into array and can iterate on that array for the compiler part.
> 
> the whole tokenizer will not be longer than 10 lines of codes.

you really should read up a bit about *recursive descent parsers* and why they do not use regular expressions most of the time.

**Comment by strager** [comment]
> you really should read up a bit about _recursive descent parsers_ and why they do not use regular expressions most of the time.

@KeremArdicli is talking about using regex for just the tokenizer/lexer, not the grammar parser.

**Comment by davidbuzz** [comment]
A 'real' php developer walk's in the room.... ;-)   This pr needs polyfills to make it work with php 7.4 and newer.

File: c.php [hunk 2]
```diff
@@ -107,18 +101,18 @@ function drop_line() {
...
             $this->trim_left();
         }
 
         if ($this->is_empty()) {
-            return false;
+            return null;
         }
 
-        $loc = $this->loc();
+        $loc   = $this->loc();
```
Comment (k1ng440) [inline c.php:115]: Nit: Remove unnecessary space

File: c.php [hunk 3]
```diff
@@ -128,17 +122,18 @@ function next_token() {
...
-            "}" => TOKEN_CCURLY,
-            "," => TOKEN_COMMA,
-            ";" => TOKEN_SEMICOLON,
-        );
+        $literal_tokens = [
+            "(" => TokenType::OPAREN,
+            ")" => TokenType::CPAREN,
+            "{" => TokenType::OCURLY,
+            "}" => TokenType::CCURLY,
+			"," => TokenType::COMMA,
```
Comment (k1ng440) [inline c.php:133]: Nit: Fix ident.

File: c.php [hunk 4]
```diff
@@ -128,17 +122,18 @@ function next_token() {
...
+            ";" => TokenType::SEMICOLON,
+		];
```
Comment (k1ng440) [inline c.php:135]: Nit: Fix ident.

File: c.php [hunk 5]
```diff
@@ -168,52 +163,48 @@ function next_token() {
...
+            return new Token($loc, TokenType::NUMBER, $value);
         }
 
         todo("next_token");
     }
 }
 
-define("TYPE_INT", "TYPE_INT");
+enum Type {
+	case INT;
```
Comment (k1ng440) [inline c.php:174]: Nit: Fix ident.

File: c.php [hunk 6]
```diff
@@ -168,52 +163,48 @@ function next_token() {
...
-    public $name;
-    public $args;
-
-    public function __construct($name, $args) {
-        $this->name = $name;
-        $this->args = $args;
-    }
+    public function __construct(
+		public Token $name,
+		public array $args
```
Comment (k1ng440) [inline c.php:180]: Nit: Fix ident.

File: c.php [hunk 7]
```diff
@@ -168,52 +163,48 @@ function next_token() {
...
 
 class RetStmt {
-    public $expr;
 
-    public function __construct($expr) {
-        $this->expr = $expr;
-    }
+    public function __construct(
+    	public int $expr
+	) {}
```
Comment (k1ng440) [inline c.php:188]: Nit: Fix ident.

File: c.php [hunk 8]
```diff
@@ -168,52 +163,48 @@ function next_token() {
...
-    public $body;
-
-    public function __construct($name, $body) {
-        $this->name = $name;
-        $this->body = $body;
-    }
+    public function __construct(
+		public $name,
+		public $body
+	) {}
```
Comment (k1ng440) [inline c.php:195]: Nit: Fix ident.

File: c.php [hunk 9]
```diff
@@ -222,84 +213,86 @@ function expect_token($lexer, ...$types) {
     echo sprintf("%s: ERROR: expected %s but got %s\n", 
         $lexer->loc()->display(),
         join(" or ", $types),
-        $token->type);
-    return false;
+        $token->type
+	);
+ 
+	return null;
```
Comment (k1ng440) [inline c.php:219]: Nit: Fix ident.

**Review by k1ng440 — CHANGES_REQUESTED** [review]
Looks good to me, but there are issues with the indentations in several spots.

**Review by eoussama — APPROVED** [review]
